### PR TITLE
Referrals

### PR DESCRIPTION
**Pull request by IanPhilips**
- Users get a prominently displayed share button in the contract details row
  - Clicking it will tack on ?Referrer={username}
  - when a signed out user lands on this page, the username and contract ids are saved to local storage
    - If there is no referrer query, the contract creator is saved as the referrer
  - When the user signs up, those cached items will be added to the newly created user object
  - A referral bonus (via txn) and notification will be sent to the referrer.
- an invite to group button tacks on the referrer query to the url and if someone signs in with that url, they'll be automatically added to that group and a referral bonus will be sent with the generic message 'joined because of you'
- The referral bonus is set at M$500 for now

Screen Shot 2022-06-28 at 6 37 56 PM

Screen Shot 2022-06-28 at 6 37 07 PM

Screen Shot 2022-06-28 at 6 37 25 PM

### CODE REVIEW

**Comment by IanPhilips** [comment]
I don't think this looks great, would love some ideas for where to place the invite to group button.
Screen Shot 2022-06-28 at 7 29 57 PM

**Comment by jahooma** [comment]
Not great layout haha

Screen Shot 2022-06-29 at 12 28 05 PM

VS.

Screen Shot 2022-06-29 at 12 29 18 PM

**Comment by jahooma** [comment]
Invite text goes off the screen on mobile:

Screen Shot 2022-06-29 at 12 31 48 PM

Maybe that invite widget should just be on the next line, below "Membership Open". Even on desktop.

File: functions/src/on-update-user.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      }
+
+      const txnDoc = await firestore.collection(`txns/`).doc(txn.id)
+      await transaction.set(txnDoc, txn)
+      console.log('created referral with txn id:', txn.id)
+      // We're currently not subtracting M$ from the house, not sure if we want to for accounting purposes.
+      transaction.update(referredByUserDoc, {
+        balance: referredByUser.balance + REFERRAL_AMOUNT,
+        totalDeposits: referredByUser.totalDeposits + REFERRAL_AMOUNT,
+      })
```
Comment (jahooma) [inline functions/src/on-update-user.ts:84]: Rather than create a txn and execute it yourself, you should call `runTxn()` from `transact.ts`.

**Review by jahooma — APPROVED** [review]
LGTM! This is great, let's ship it! Thanks, Ian.

File: functions/src/on-update-user.ts (hunk 1 above)
Comment (IanPhilips) [reply functions/src/on-update-user.ts:84]: I actually tried this but this doesn't work if I want the txn to be tied to an idempotent event. I could create my own runTxn with the requirements (passing the txn id) for my function if you'd like.

File: functions/src/on-update-user.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+export const ReferredUserDescriptionPrefix = 'Referred user id'
+
+export const onUpdateUser = functions.firestore
+  .document('users/{userId}')
+  .onUpdate(async (change, context) => {
+    const prevUser = change.before.data() as User
+    const user = change.after.data() as User
+    const { eventId } = context
+
+    if (prevUser.referredByUserId === user.referredByUserId) {
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/on-update-user.ts:20]: Good idea!

File: functions/src/on-update-user.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    })
+    // If user has a referral txn contains 'referred by' the current username in the description, halt
+    if (referralTxns.map((txn) => txn.description).includes(referredByUserId)) {
+      console.log('found referral txn with the same details, aborting')
+      return
+    }
+    console.log('creating referral txns')
+    const fromId = HOUSE_LIQUIDITY_PROVIDER_ID
+
+    await firestore.runTransaction(async (transaction) => {
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply functions/src/on-update-user.ts:62]: Good to know, will adapt em

File: web/components/referrals-button.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+function ReferralsDialog(props: {
+  user: User
+  referralIds: string[]
+  isOpen: boolean
+  setIsOpen: (isOpen: boolean) => void
+}) {
+  const { user, referralIds, isOpen, setIsOpen } = props
+
+  useEffect(() => {
+    prefetchUsers([...referralIds])
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/components/referrals-button.tsx:45]: good eye!

File: web/lib/firebase/groups.ts [hunk 5]
```diff
@@ -95,6 +95,15 @@ export async function getGroupsWithContractId(
   setGroups(groups)
 }
 
+export async function addUserToGroupViaSlug(groupSlug: string, userId: string) {
+  // get group to get the member ids
+  const group = await getGroupBySlug(groupSlug)
+  if (group && !group.memberIds.includes(userId))
+    return await updateGroup(group, {
+      memberIds: [userId, ...group.memberIds],
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/lib/firebase/groups.ts:103]: good idea!

File: web/lib/firebase/users.ts [hunk 6]
```diff
@@ -87,12 +90,83 @@ export function listenForPrivateUser(
...
 
 const warmUpCreateUser = throttle(createUser, 5000 /* ms */)
 
+export function writeReferralInfo(
+  contractUsername: string,
+  contractId?: string,
+  referralUsername?: string,
+  groupSlug?: string
+) {
+  const local = safeLocalStorage()
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/lib/firebase/users.ts:108]: It looks like this is basically local storage:
`export const safeLocalStorage = () => (isLocalStorage() ? localStorage : null)`

File: web/pages/group/[...slugs]/index.tsx [hunk 7]
```diff
@@ -157,6 +164,12 @@ export default function GroupPage(props: {
   const isCreator = user && group && user.id === group.creatorId
   const isMember = user && memberIds.includes(user.id)
 
+  const { referrer } = router.query as {
+    referrer?: string
+  }
+  if (!user && router.isReady)
+    writeReferralInfo(creator.username, undefined, referrer, group.slug)
```
[the remark this answers is not in the document]
Comment (IanPhilips) [reply web/pages/group/[...slugs]/index.tsx:171]: Sounds good

**Pull request merged**